embedded/store/txpool.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2022 CodeNotary, Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package store
+
+import (
+	"sync"
+
+	"github.com/codenotary/immudb/embedded/htree"
+)
+
+type txPoolOptions struct {
+	PoolSize     int
+	MaxTxEntries int
+	MaxKeyLen    int
+	Preallocated bool
+}
+
+type TxPool interface {
+	Alloc() (*Tx, error)
+	Release(*Tx)
+	Stats() (used, free, max int)
+}
+
+type txPool struct {
+	pool []*Tx
+	used int
+	m    sync.Mutex
+	opts txPoolOptions
+}
+
+func newTxPool(opts txPoolOptions) (TxPool, error) {
+
+	if opts.PoolSize <= 0 || opts.MaxTxEntries <= 0 || opts.MaxKeyLen <= 0 {
+		return nil, ErrIllegalArguments
+	}
+
+	ret := &txPool{
+		pool: make([]*Tx, 0, opts.PoolSize),
+		opts: opts,
+	}
+
+	if opts.Preallocated {
+		// The pool uses only 5 allocations in total
+		// instead of allocating data separately for each tx and then each
+		// entry, we instead allocate single large arrays large enough
+		// to fulfill requirements of all internal objects
+		//
+		// Keeping the number of allocated objects small is essential
+		// for reducing the CPU time spent in GC cycles.
+
+		txBuffer := make([]Tx, opts.PoolSize)
+		txEntryPtrBuffer := make([]*TxEntry, opts.PoolSize*opts.MaxTxEntries)
+		txEntryBuffer := make([]TxEntry, opts.PoolSize*opts.MaxTxEntries)
+		headerBuffer := make([]TxHeader, opts.PoolSize)
+		keyBuffer := make([]byte, opts.PoolSize*opts.MaxTxEntries*opts.MaxKeyLen)
+
+		for i := 0; i < opts.PoolSize; i++ {
+			tx := &txBuffer[i]
+			ret.pool = append(ret.pool, tx)
+
+			tx.entries = txEntryPtrBuffer[:opts.MaxTxEntries]
```

Comment:
aren't all txs getting the same slice?

Reply:
oh, final lines are doing the trick... `txEntryPtrBuffer = txEntryPtrBuffer[opts.MaxTxEntries:]`

Reply:
was expecting `i` to participate, it may be easier to read as it doesn't require to read all the lines

Reply:
does it makes the difference to white-box initialize the txholders here compared of using newTx method? which it's already alloc optimized with a similar approach?

Reply:
another point to consider, this shared array may lead to concurrency issues... as the requirement for this to be consistent, the array must be threadsafe...

Reply:
regarding, this is thread safety - we don't change the allocation or slices once the initialization is done. Elements of the same array can also be written to concurrently (good explanation is here: https://stackoverflow.com/a/49879469/18309616) - as long as each goroutine works on its own part of the array, it is safe. I was confused by this some time ago too - I thought that it would be thread safe as long as each goroutine works on a separate cache line in CPU but it seems like CPUs have a way around this.

Reply:
Regarding one large allocation - I'll check that. I agree that it may not bring any benefit, then I'll switch to allocating each txholder separately.

Reply:
I did performance comparision between two implementations, there are no differences discovered thus I've switched to separate allocation for each entry in 03dcbcc1.

![Zrzut ekranu z 2022-07-29 11-12-01](https://user-images.githubusercontent.com/1365626/181726597-a1f5a1b0-3d03-4177-964e-36f72ca21f4a.png)